src/librustc_resolve/late.rs

```diff
@@ -736,14 +736,20 @@ impl<'a, 'b, 'ast> LateResolutionVisitor<'a, 'b, 'ast> {
     /// Searches the current set of local scopes for labels. Returns the first non-`None` label that
     /// is returned by the given predicate function
     ///
-    /// Stops after meeting a closure.
-    fn search_label<P, R>(&self, mut ident: Ident, pred: P) -> Option<R>
+    /// Stops after meeting a closure (unless `passthrough_closure` is `true` - for diagnostic use).
+    fn search_label<P, R>(
+        &self,
+        mut ident: Ident,
+        passthrough_closure: bool,
```

Comment:
Well, I think this should be done differently.
See how the next cases are treated
```
fn main() {
    let a = 10;
    fn foo() {
        a; // ERROR can't capture dynamic environment in a fn item
    }
    [0; a]; // ERROR attempt to use a non-constant value in a constant
}
```
We *always* walk all the scopes ignoring "barrier" ribs and then run `validate_res_from_ribs` to report errors if we crossed some barriers that we cannot cross.
(This also deserves a separate `ResolutionError` variant because this is not an "unresolved label" error, we did find the resolution, it was just behind a barrier.)

(I also dislike how typo candidates are found here, but that's a separate story.)

Reply:
I've looked at how `validate_res_from_ribs` is being used and how it differs from how `search_label` is working - I can see a handful of different approaches to reconcile the differences, but I'm not sure which you'd prefer (or intended).

`validate_res_from_ribs` expects a `Res` - but we don't have one of these when looking for labels, just a `NodeId`. One approach (that feels like most like the example you provided) would be:

1. Modify `search_labels` to walk all scopes, ignoring "barrier" ribs, returning `ResolutionError::UnresolvedLabel` if there were no match.
2. Introduce a `Res::Label` (or something like that).
3. Extend `validate_res_from_ribs` to handle the `Res::Label` case, returning `ResolutionError::LabelOutOfScope` (or something like that) if the `Res::Label` was found through a "barrier" rib.

However, if introducing a `Res::Label` would be inappropriate, a `validate_label_from_ribs` function could be introduced, which performs the function of `validate_res_from_ribs` but for labels - `search_labels` would be modified as in the previous proposal.

Do either of these approaches describe what you had in mind, or something else?

Reply:
I don't mean reusing `validate_res_from_ribs` literally (`Res::Label` certainly shouldn't be reintroduced), just using the same general approach of looking through all the scopes and detecting bad resolutions in a post-processing step.

Reply:
I've made a bunch of changes in the most recent commit which I think is closer to what you were wanting, but might not be exactly right. I tried to mirror the relationship between `validate_res_from_ribs` and `resolve_ident_in_lexical_scope` (the only caller of `validate_res_from_ribs`) but deviated a little when it helped keep diagnostic logic separate and maintain the diagnostic quality.